internal/initdata/initdata_test.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/stretchr/testify/require"
+)
+
+func buildTemplate(algorithm, version string, data bool) string {
+	var builder strings.Builder
+
+	if algorithm != "" {
+		fmt.Fprintf(&builder, "algorithm = \"%s\"\n", algorithm)
+	}
+	if version != "" {
+		fmt.Fprintf(&builder, "version = \"%s\"\n", version)
+	}
+
+	if data {
+		builder.WriteString(`
+[data]
+somekey = "somevalue"
+"imagepuller.toml" = '''
+socket = 'unix:///run/guest-services/imagepuller.sock'
+credentials = []
+'''
+`)
+	}
+	return builder.String()
+}
+
+func tomlToAnnotation(toml string) string {
+	compressed, _ := compress([]byte(toml))
+	encoded := base64.StdEncoding.EncodeToString(compressed)
+	return encoded
+}
+
+func annotationToToml(annotation string) string {
+	decoded, _ := base64.StdEncoding.DecodeString(annotation)
+	decompressed, _ := decompress(decoded)
+	return string(decompressed)
+}
+
+func TestDecodeKataAnnotation(t *testing.T) {
+	tests := map[string]struct {
+		version   string
+		algorithm string
+		data      bool
+		wantErr   error
+	}{
+		"success sha256":       {version: InitdataVersion, algorithm: "sha256", data: true},
+		"success sha-256":      {version: InitdataVersion, algorithm: "sha256", data: true},
+		"success sha384":       {version: InitdataVersion, algorithm: "sha384", data: true},
+		"success sha-384":      {version: InitdataVersion, algorithm: "sha384", data: true},
+		"success sha512":       {version: InitdataVersion, algorithm: "sha512", data: true},
+		"success sha-512":      {version: InitdataVersion, algorithm: "sha512", data: true},
+		"version missing":      {version: "", wantErr: errVersionMismatch},
+		"version unknown":      {version: "1.1.0", wantErr: errVersionMismatch},
+		"algorithm missing":    {version: InitdataVersion, wantErr: errAlgorithmUnknown},
+		"algorithm unknown":    {version: InitdataVersion, wantErr: errAlgorithmUnknown},
+		"success data missing": {version: InitdataVersion, algorithm: "sha256", data: false},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			assert := assert.New(t)
+			require := require.New(t)
+
+			annotation := tomlToAnnotation(buildTemplate(tc.algorithm, tc.version, tc.data))
+			i, err := DecodeKataAnnotation(annotation)
+			if tc.wantErr != nil {
+				require.ErrorIs(err, tc.wantErr)
+				return
+			}
+
+			require.NoError(err)
+			assert.Equal(tc.algorithm, i.Algorithm)
+			assert.Equal(InitdataVersion, i.Version)
+
+			if tc.data {
+				assert.Contains(i.Data, "somekey")
+				assert.Contains(i.Data, "imagepuller.toml")
+			}
+		})
+	}
+}
+
+func TestEncodeKataAnnotation(t *testing.T) {
+	tests := map[string]struct {
+		version   string
+		algorithm string
+		data      map[string]string
+		want      string
+		wantErr   error
+	}{
+		"success sha256": {
+			version:   InitdataVersion,
+			algorithm: "sha256",
+			data:      map[string]string{"key": "value"},
+			want:      "a1c4b2afd1fa1c06bab23f0e7d8b6bbc07e25a3d0240e79f0048da0775bae6a5",
```

Comment:
This is now testing the stability of toml encoding. Maybe a roundtrip test `Decode(Encode(foo)) == foo` would be more robust.